### DISCUSSION TITLE
Issue with I2C oled display

### DISCUSSION THREAD

**Discussion by Carlito06150** [pschatzmann/arduino-audio-tools#1728] [Q&A] [answered]
Hello,

I am using this type of code in setup:
AudioLogger::instance().begin(Serial, AudioLogger::Warning);
auto config = kit.defaultConfig(TX_MODE); 
config.sd_active = true;
kit.begin(config);
SD.begin(chipSelect);
audioFile = SD.open("/217007.mp3");
decoder.begin();
copier.setCheckAvailableForWrite(false);
copier.begin(decoder, audioFile);

no pb, but I decided to add an oled display using this code:
#include <Wire.h>
#include <Adafruit_GFX.h>
#include <Adafruit_SSD1306.h>
#define nombreDePixelsEnLargeur 128         
#define nombreDePixelsEnHauteur 64         
#define brocheResetOLED         -1         
#define adresseI2CecranOLED     0x3C       
#define portDisplay1            18
#define portDisplay2            23

Adafruit_SSD1306 oled(nombreDePixelsEnLargeur, nombreDePixelsEnHauteur, &Wire, brocheResetOLED);

and in setup:
Wire.begin(portDisplay1,portDisplay2);
  Serial.begin(115200);
  // SSD1306_SWITCHCAPVCC = generate display voltage from 3.3V internally
  if(!oled.begin(SSD1306_SWITCHCAPVCC, 0x3C)) { 
    Serial.println(F("SSD1306 allocation failed"));
    for(;;); // Don't proceed, loop forever
  }
oled.clearDisplay();
  testdrawline();

Separately no pb, but if I have these two pieces of code, the oled does nothing if I call kit.begin(config);...

Any idea?

**Comment by pschatzmann** [comment, marked as answer]
The AudioKit already uses Wire to communicate with the codec using the pins that are relevant for the Codec!
Use Wire1 for the display or force the AudioKit to use Wire1 instead...

**Reply by Carlito06150** [reply]
oooooh thx a lot for this quick answer!!!

I'am trying now :p

**Reply by Carlito06150** [reply]
but no change :( 
Adafruit_SSD1306 oled(nombreDePixelsEnLargeur, nombreDePixelsEnHauteur, &Wire1, brocheResetOLED); at the top and in the setup:
Wire1.begin(portDisplay1,portDisplay2);

**Reply by Carlito06150** [reply]
also tried with another library SSD1306, same result... :/

**Reply by Carlito06150** [reply]
could it be linked to the ports I use? (18 & 23)

**Reply by pschatzmann** [reply]
I assume you use an AudioKit board: so Look up the used pins with the help of your used board in the [audio-driver](https://github.com/pschatzmann/arduino-audio-driver) project!

**Reply by Carlito06150** [reply]
I assume you use an AudioKit board
--> you assume well :) a 8388 (https://fr.aliexpress.com/item/4000130915903.html?spm=a2g0o.order_list.order_list_main.23.41025e5bsULoFN&gatewayAdapt=glo2fra)
so Look up the used pins with the help of your used board in the audio-driver project
--> hummm, not sure to understand, could you precise? (sorry to annoy you, all this is a new world for me, I managed to do what I want with your great library and neogps one, but it begins to be too difficult for me :( )


in the definition I can see that:
addPin(PinFunction::KEY, 23, PinLogic::Input, 4);
addPin(PinFunction::KEY, 18, PinLogic::Input, 5);

what's the link between that and the use of 18 & 23 pins?

I also see 
addPin(PinFunction::PA, 21, PinLogic::Output);
addPin(PinFunction::LED, 22, PinLogic::Output);
and I use 21 & 22 IO for my gps module without issue so it must not be the pb...

**Reply by pschatzmann** [reply]
Assuming you are using  AudioKitEs8388V1 search for it and you will find this line

    static AudioBoard AudioKitEs8388V1{AudioDriverES8388, PinsAudioKitEs8388v1};

So you want to look up PinsAudioKitEs8388v1 which gives

    static PinsAudioKitEs8388v1Class PinsAudioKitEs8388v1;

And looking up PinsAudioKitEs8388v1Class gives you the following that contains all pins that are used somehow by this baord

```C++

class PinsAudioKitEs8388v1Class : public DriverPins {
 public:
  PinsAudioKitEs8388v1Class() {
    // sd pins
    addSPI(ESP32PinsSD);
    // add i2c codec pins: scl, sda, port, frequency
    addI2C(PinFunction::CODEC, 32, 33);
    // add i2s pins: mclk, bck, ws,data_out, data_in ,(port)
    addI2S(PinFunction::CODEC, 0, 27, 25, 26, 35);

    // add other pins
    addPin(PinFunction::KEY, 36, PinLogic::InputActiveLow, 1);
    addPin(PinFunction::KEY, 13, PinLogic::InputActiveLow, 2);
    addPin(PinFunction::KEY, 19, PinLogic::InputActiveLow, 3);
    addPin(PinFunction::KEY, 23, PinLogic::InputActiveLow, 4);
    addPin(PinFunction::KEY, 18, PinLogic::InputActiveLow, 5);
    addPin(PinFunction::KEY, 5, PinLogic::InputActiveLow, 6);
    addPin(PinFunction::AUXIN_DETECT, 12, PinLogic::InputActiveLow);
    addPin(PinFunction::HEADPHONE_DETECT, 39, PinLogic::InputActiveLow);
    addPin(PinFunction::PA, 21, PinLogic::Output);
    addPin(PinFunction::LED, 22, PinLogic::Output);
  }
};

```
GPIO18 is used by the key 5 and 23 by key 4 !

**Reply by Carlito06150** [reply]
that's exactly what I found in my previous message (I am not too bad :p )
does it mean it's dead for me or I can just remove these 2 lines?

what's the meaning of "PA" at the end? I'm using GPIO 21&22 for my gps module without any issue...

**Reply by pschatzmann** [reply]
These pins are hard wired on the board and should therefore not be used. E.g if you press the key 5 you mess up the signal that you generate for the screen. 
The PA is the power amplifier and you should not use this as well: It activates/deactivates the amplifier that does the output to the speaker output pins.

**Reply by Carlito06150** [reply]
arg :( well in fact I don't use the buttons, but ok... are there 4 GPIOS I can use somewhere? I need 2 for my gps module and 2 others for the screen... 
on the board I can see IO5, IO0, TX0 and RX0...

**Reply by pschatzmann** [reply]
IO0 is the I2S master clock.
Maybe you can also use the pins on the debug port...

**Comment by Carlito06150** [comment]
debug port? 
![audiokit](https://github.com/user-attachments/assets/3e89dfab-0cb7-44f3-a305-73812d1a689f)

**Reply by pschatzmann** [reply]
If you don't use the SD card you can also use 13 and 15 by setting the dip

PinsSPI ESP32PinsSD{PinFunction::SD, 14, 2, 15, 13, SPI};

**Reply by Carlito06150** [reply]
bad news: I need the SD Card, audio files and data are stored on it... :|

**Reply by pschatzmann** [reply]
Not sure if the pin keys are usable for your purpose: Did you try them w/o the AudioTools ? 
If so you can try to create your own board  or deactivate the keys via the API, so that they are not setup up automatically.

**Reply by Carlito06150** [reply]
in fact we come back to my initial message in this post? without the audiotools, on GPIOS 18 and 23, it works and with the audiotools, from the point I call kit.begin(config); the screen does not react to the commands...

**Reply by pschatzmann** [reply]
So, just change the sequence and setup the kit first. 
But it would be saver to do what I described above.

**Reply by Carlito06150** [reply]
I try I try, I will see this evening back to home if I can find a correct sequence

**Reply by Carlito06150** [reply]
arf, I've just put the screen code at the end of setup and.. it works!!!
but... not totally
in my code I have this: kit.addAction(kit.getKey(5), volDown);
and everytime I write something on the screen, the volDown function is called :D

**Reply by Carlito06150** [reply]
for the moment I just comment these lines: 
//addPin(PinFunction::KEY, 23, PinLogic::InputActiveLow, 4);
//addPin(PinFunction::KEY, 18, PinLogic::InputActiveLow, 5);
at least I can go on in the rest of my projet...

**Reply by pschatzmann** [reply]
you better deactivate them in your sketch like this before calling begin:
    
    auto pin = kit.board().getPins().getPin(PinFunction::KEY,4);
    if (pin) pin.value().pin_logic = PinLogic::Inactive;


Prefer to use the I2SCodecStream which does not implement any key functions

**Reply by Carlito06150** [reply]
ok I try

**Reply by Carlito06150** [reply]
compiler doesn't like :/

**Reply by Carlito06150** [reply]
> Prefer to use the I2SCodecStream which does not implement any key functions

well I will search how to use it, I don't know how to change my code to do that, I gonna check your examples

**Reply by Carlito06150** [reply]
hummm one last remaining issue... each time I display something, the music that is played is badly impacted... would you have an idea to find the reason why?

**Reply by pschatzmann** [reply]
Did you read [this Wiki](https://github.com/pschatzmann/arduino-audio-tools/wiki/It's-not-working) ? You should be able to figure this out yourslef!

**Reply by Carlito06150** [reply]
Well, I managed to have it in a separate task, same pb, so it must be a pb of resource... I go on...

**Reply by pschatzmann** [reply]
Maybe you are still starving the audio task: try to add some small delay() to your output task to render control to the audio task.

**Comment by Carlito06150** [comment]
hummm

in my setup:
  SD.begin(chipSelect);
  audioFile = SD.open("/217007.mp3");
  decoder.begin();
  copier.setCheckAvailableForWrite(false);
xTaskCreatePinnedToCore(TaskPlayerFunc,"TaskPlayer", 10000,NULL,10,&TaskPlayer, 1);   

my method:
void TaskPlayerFunc( void * pvParameters ){
  for(;;){
    Serial.printf("%s running on core %d (priorite %d) - %dms - fichier %s\n", "TaskPlayer", xPortGetCoreID(), uxTaskPriorityGet( NULL ),millis()-previousTimePlayer,audioFile.name());
    previousTimePlayer=millis();
    copier.begin(decoder, audioFile);
    vTaskDelay(200/ portTICK_PERIOD_MS);//vTaskDelay( pdMS_TO_TICKS( 200 ) );
  }
  vTaskDelete(NULL);
}

and in the loop, each time if display something on the screen, the sound is bugged...

**Reply by pschatzmann** [reply]
As I was writing above your issue is most likey in the LED dispay functionality which you did not share!
Finally don't add any delays and prints in the Audio loop: This has been clearly stated im my Wiki!

Have a look at the [basic example](https://github.com/pschatzmann/arduino-audio-tools/wiki/Multicore-Processing#audio-tasks-a-basic-example) in the [Multicore Processing](https://github.com/pschatzmann/arduino-audio-tools/wiki/Multicore-Processing) Wiki: the copy should be in the loop, not the setup of the copy!

If this alone did not help, try to add a delay(1) after every display function that triggers a Wire transaction! This is to make sure that you don't starve the audio task!

**Reply by Carlito06150** [reply]
well copy is also in the loop... in the setup it's like a startup song and after in the loop I load another file depending on my position given by the gps module...

no delays in the loop as I read in the wiki...

the oled acces is something like that:
oled.setTextColor(WHITE,BLACK);
oled.setTextSize(1);
oled.setCursor(0, 46);
oled.println("timer:             ");
oled.setCursor(0, 46);
oled.println("timer:"+String(t));
oled.display();

oled is intialised like that in setup:
Wire1.begin(portDisplay1,portDisplay2);
if(!oled.begin(SSD1306_SWITCHCAPVCC, 0x3C)) { 
    Serial.println(F("SSD1306 allocation failed"));
    for(;;); // Don't proceed, loop forever
}

**Reply by pschatzmann** [reply]
Copy needs to be in a separate task!
And did you try to add some delay before or after oled.display() that I assume is in the loop task ?
I am not sure if the other commands execute a Wire transaction...

Please have a look at the example in the link that I have sent you. There is not point in me repeating the same again!

**Reply by Carlito06150** [reply]
thanks a lot! it works with Audio Tasks: A Basic Example in Multicore Processing :)

last question, in my setup I have this code:
SD.begin(chipSelect);
audioFile = SD.open("/217007.mp3");
decoder.begin();
copier.setCheckAvailableForWrite(false);
copier.begin(decoder, audioFile);      
task.begin([](){copier.copy();}); 

If I want in the loop to load another mp3 song, what's the best way to do it? can I do it with the same task, do I need to kill it and create another one?

But thx a lot for your help, your patience and your project :)

**Comment by pschatzmann** [comment]
You can remove the copier.setCheckAvailableForWrite(false); from your sketch since this is false by default

It might be a good idea to suspend and resume the task. 
I would try the following to process another file

```C++
void processFile(const char *name){
    task.suspend();
    audioFile.close();
    audioFile = SD.open(name);
    decoder.begin();  
    copier.begin(decoder, audioFile);
    task.resume();
}
```

**Reply by Carlito06150** [reply]
works perfectly ;)

it was working before but I was killing the task, a bit to strong, resuming is better :)

the las of the last: everytime I call SD.open... it seems there is a little memory leak, are you aware of this?

**Reply by pschatzmann** [reply]
Don't forget to close the file: I updated the logic above...

**Reply by Carlito06150** [reply]
you are my god :)

**Reply by waszee** [reply]
Carlito I have been tryiing something similar to you with the AI Thinker audio board.  I am using the player to play either from a list of mp3 located in a file on the SD card or from a list of URL addresses located in anotther file on the SD card to play internet stations.  I modified the button controls to switch stations or between mp3 and internet and out put some information to the Serrial Monitor.  I am currently outputting metadata to the serial monitor using the printMetaData but I am sure that code could easily be modify to out to a SS1306 display.  I already do that in another version that uses audio  5102  module with with esp32s3-wroom1 dev kit board.  Unlike the AI Thinker the dev kit has plenty of free gpio pins. What pins did you use for the display on your board?  Below shows the basic playing logic...  Hope this gives you some ideas.  I need to now go bug Phil for som e help handling failed URL connection errors...

"""
  // setup playerAudioSource source = source1 mp3, source2 url addr;
  if (mp3play==true) {
    player.setAudioSource(source1);
    max_index=max_mp3; }
  else  {
    player.setAudioSource(source2);
    max_index=max_urls;
  }
  player.setOutput(kit);
  player.setDecoder(decoder);
  player.setMetadataCallback(printMetaData,SELECT_ANY);
  player.setVolume(vol);
  player.begin();
}

// Arduino loop
void loop() {
  player.copy();
  kit.processActions();
  
}
"""

**Comment by Carlito06150** [comment]
18 & 23 with this code for 4 & 5:
auto pin = kit.board().getPins().getPin(PinFunction::KEY,4);
if (pin) pin.value().pin_logic = PinLogic::Inactive;

**Reply by waszee** [reply]
Thank you.  I guess you had to declare a second i2c port.  Thought it weird that AI Thinker did not provide the pin outs to the I2C  used by the audio chip since those commands are just sent once and have little load.  FYI having some trouble with the IR remote getting the stations to change.  I think there may be a bug in the player.setIndex(idx) command when an i2s audio output is used.  That command worked on the Thinker Audiokit using the buttons to trigger, but on the I2s system seems to just try to move the metadata output to new station but not move the audio connection.  Well that is this morning project.  Problem could easily be a simple mistake in my code..

**Reply by Carlito06150** [reply]
hummm declare a second I2C port? no...

**Comment by TangerineDreamer** [comment]
Hey,have you ever tried audiokit with a big TFT and touch?
I can also manage a 4" TFT without changing anything in the audiokit,but with no TFT touch,only using the audio kit buttons.

https://github.com/pschatzmann/arduino-audiokit/discussions/60
